### ISSUE TITLE
[SW] Crash or abnormal behavior when setting the resolution to 2x

### ISSUE THREAD

**Issue by TransparentBlue** [libretro/beetle-psx-libretro#211]
RetroArch 1.6.3 stable
Windows 7x64
Core Version:  aefcac9
RetroArch Driver: Vulkan and GL
Software renderer

If the settings is changed while a game is loaded retroarch hangs until you force quit it through the task manager
If set beforehand it will either crash and exit on it's own shortly after the BIOS sound starts being heard, and occasionally will start but only showing the upper left corner of the game.

**Comment by twinaphex** [comment]
I can reproduce this, and it seems to go back a long time as well.

**Comment by rz5** [comment]
Same here, here's the full backtrace:

```
Thread 1 received signal SIGSEGV, Segmentation fault.
0x00007ff80c8438c7 in msvcrt!memset () from C:\WINDOWS\System32\msvcrt.dll
(gdb) bt full
#0  0x00007ff80c8438c7 in msvcrt!memset () from C:\WINDOWS\System32\msvcrt.dll
No symbol table info available.
#1  0x0000000014b67b8d in GPU_Update (sys_timestamp=309967)
    at mednafen/psx/gpu.cpp:1655
        src = 0x34151518
        i = 0
        y = 486
        udx_end = 1340
        _upscale = 2
        x = 79885056
        udmw = 1400
        udx_start = 60
        ufb_x = 0
        fb_x = 0
        dx_start = 30
        dx_end = 670
        dest_line = 241
        FirstVisibleLine = 16
        displayfb_yoffset = 241
        dmw_width = 0
        pix_clock_offset = 0
        dest = 0xecd1820
        VisibleLineCount = 240
        pix_clock = 0
        pix_clock_div = 0
        chunk_clocks = 49
        dot_clocks = 13
        gpu_clocks = 0
        gpu = 0x33f68040
        DotClockRatios = {10, 8, 5, 4, 7}
        dmc = 3
        dmw = 700
        sys_clocks = 31
        next_dt = 0
#2  0x0000000014d5cb15 in PSX_EventHandler (timestamp=309967)
    at libretro.cpp:419
        nt = 310064
        prev = 0x15055aa0 <events>
        e = 0x15055ab8 <events+24>
#3  0x0000000014dcc3e7 in PS_CPU::RunReal<false> (this=0x7e637a0,
    timestamp_in=0) at mednafen/psx/cpu.cpp:2428
        PC = 2147851772
        new_PC = 4
        new_PC_mask = 4294967295
        LDWhich = 2
        LDValue = 27707
        timestamp = 309967
#4  0x0000000014b3aae7 in PS_CPU::Run (this=0x7e637a0, timestamp_in=0)
    at mednafen/psx/cpu.cpp:2447
No locals.
#5  0x0000000014d64d41 in retro_run () at libretro.cpp:3701
        updated = true
        rects = {2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2, 700, 2,
          700, 2, 700, 2, 700, 2, 700, 2, 700...}
        spec = {surface = 0x55c06f0, VideoFormatChanged = false,
          DisplayRect = {x = 0, y = 0, w = 0, h = 480},
          LineWidths = 0x15055d40 <retro_run::rects>, IsFMV = 0x0,
          InterlaceOn = true, InterlaceField = true, skip = 0,
          SoundFormatChanged = false, SoundRate = 44100, SoundBuf = 0x0,
          SoundBufMaxSize = 0, SoundBufSize = 0, SoundBufSizeALMS = 0,
          MasterCycles = 0, MasterCyclesALMS = 0, SoundVolume = 1,
          soundmultiplier = 1, NeedRewind = false, NeedSoundReverse = false}
        espec = 0x4c2fb40
        timestamp = 0
        fb = 0x1dc
        width = 640
        height = 480
        upscale_shift = 0 '\000'
        interbuf = 0x14e4b480 <IntermediateBuffer>
#6  0x0000000000402e7d in core_run () at core_impl.c:415
No locals.
#7  0x0000000000408519 in runloop_iterate (sleep_ms=0x4c2fdd0)
    at retroarch.c:3054
        i = 5
        input_nonblock_state = false
        settings = 0x553b4a0
        max_users = 5
#8  0x000000000040175c in rarch_main (argc=1, argv=0x1dbbc0, data=0x0)
    at frontend/frontend.c:131
        sleep_ms = 0
        ret = 0
        args = 0x0
#9  0x00000000004017bc in SDL_main (argc=1, argv=0x1dbbc0)
    at frontend/frontend.c:151
No locals.
#10 0x00000000006f2e38 in main_getcmdline ()
No symbol table info available.
#11 0x00000000004013f8 in __tmainCRTStartup ()
    at C:/repo/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:334
        lock_free = <optimized out>
        fiberid = <optimized out>
        nested = <optimized out>
        lpszCommandLine = <optimized out>
        StartupInfo = {cb = 104, lpReserved = 0x4e1bf10 "",
          lpDesktop = 0x4e0e140 "Winsta0\\Default",
          lpTitle = 0x4e1c1e0 "C:\\Users\\razev\\Documents\\RA\\retroarch_debug.exe", dwX = 0, dwY = 0, dwXSize = 0, dwYSize = 0, dwXCountChars = 0,
          dwYCountChars = 0, dwFillAttribute = 0, dwFlags = 0,
          wShowWindow = 0, cbReserved2 = 0, lpReserved2 = 0x0,
          hStdInput = 0xffffffffffffffff, hStdOutput = 0xffffffffffffffff,
          hStdError = 0xffffffffffffffff}
        inDoubleQuote = <optimized out>
#12 0x00000000004014eb in WinMainCRTStartup ()
    at C:/repo/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:184
        ret = 255

```

**Comment by rz5** [comment]
First bad commit that causes this bug:
https://github.com/libretro/beetle-psx-libretro/commit/7bf0144c2a98ffcba5604dca47df2d4c0ca2893b

**Comment by rz5** [comment]
@TransparentBlue If you can build from source, build the latest commit and see if it fixes the issues you found.
Or else wait for the buildbot to spit out a build.

If you can, also test out the dither mask upscaling core option, I flipped the logic around (NATIVE, we do 0 shifts; UPSCALED, we do the same shifts as on the resolution). Not sure if it was correct previously or of it is right now. @simias - thoughts on this?

**Comment by TransparentBlue** [comment]
@rz5 Sorry for the delay, using the latest build currently on the buildbot 34321f5 the issue is partially fixed. 
Upon changing the resolution the game will resume but most of the graphics are gone, making and loading a savestate won't fix the issue, but loading any savestate made prior to changing the resolution will make the game resume correctly at the selected resolution.

**Comment by twinaphex** [comment]
Yeah the restoring of the state is an issue, and we still don't quite know what the culprit inside the problematic commit that was at the root of all this. We fixed the restoring of the state for the GL renderer, but software renderer is still an issue unfortunately.

**Comment by rz5** [comment]
Alright, an update on this issue.

The first commit that causes the software renderer to having bad/missing graphics when we upscale is https://github.com/libretro/beetle-psx-libretro/commit/7bf0144c2a98ffcba5604dca47df2d4c0ca2893b

Specifically, that's when @twinaphex backports the changes made in Mednafen 0.9.44.1 (https://github.com/libretro-mirrors/mednafen-(git/commit/6fe724a8b82c5d2252a39718ad88b3ea36f6e1cb), the upstream changelog says the following:
```
April 19, 2017:
    PSX: Restructured the GPU emulation code and data a bit to get a minor performance boost.
```
-------------
I have been eyeballing the changes real hard, to see anything that looks off or something, but changes that Twinaphex did look to me like they'll have the same side-effects (obviously it doesn't, given the bug).

The easy solution is to just rollback the changes of that commit and upscaling will again work without a need to restart. I'd rather fix the actual bug and have the alleged 'minor performance boost' but haven't been having much success while debugging it.

Suggestions are welcome.

**Comment by rz5** [comment]
The issue should be fixed with the latest changes.

For posterity's sake, I'll say that the bug probably stemmed from the fact that, when we told the GPU to upscale itself, the GPU never actually saved the upscaling factor inside itself during the whole thing.

This is a crucial step because there are at least 3 macros and 1 function that check what the current upscaling factor is. If we don't do the above mentioned save before calling those macros/functions, things won't work out too good.

@TransparentBlue - Could you test it and report back?

**Comment by TransparentBlue** [comment]
Yeah, it's fixed completely.

**Comment by simias** [comment]
I just stumbled on a reference to this issue while hacking on beetle's GPU code, I was surprised to find this:

https://github.com/libretro/beetle-psx-libretro/blob/4a75947e0f979c10ab0337d9bbd7b4f485674537/mednafen/psx/gpu.h#L216-L220

The reason I did the variable array trickery was because when I first added the pointer indirection for the VRAM in the upscaling patch I was told that there was a performance hit, so I thought I'd just concatenate the vram to the GPU struct to always have a static offset from the GPU pointer.

If that was reverted doesn't it also mean that performance took a hit as well?

**Comment by rz5** [comment]
@simias - that was me and I admit I didn't do a before/after benchmark. In the process of fixing #211, I compared that file to what's in upstream mednafen, carefully copying what's in there to here. Some of my thoughts are written on this commit message - https://github.com/libretro/beetle-psx-libretro/commit/fad937242768e0b443cf139eea13d91a23433526 - and in the other files I touched.

IIRC, I was having a hard time working around that variable-sized struct trickery and at the same time having our fork match what's in upstream, so I dumbed it down for my sake and sanity.
